src/storage/ExtensionBasedMapper.ts

```diff
@@ -22,52 +24,131 @@ export interface ResourcePath {
+ * In case the extension of the identifier does not correspond to the correct content-type,
+ * a new extension will be appended (with a `$` in front of it).
+ * E.g. if the path is `input.ttl` with content-type `text/plain`, the path would actually be `input.ttl$.txt`.
+ * This new extension is stripped again when generating an identifier.
+ */
 export class ExtensionBasedMapper implements FileIdentifierMapper {
-  private readonly base: string;
-  private readonly prootFilepath: string;
+  private readonly baseRequestURI: string;
+  private readonly rootFilepath: string;
   private readonly types: Record<string, any>;
 
   public constructor(base: string, rootFilepath: string, overrideTypes = { acl: TEXT_TURTLE, metadata: TEXT_TURTLE }) {
-    this.base = base;
-    this.prootFilepath = rootFilepath;
-    this.types = { ...types, ...overrideTypes };
-  }
-
-  public get baseRequestURI(): string {
-    return trimTrailingSlashes(this.base);
-  }
-
-  public get rootFilepath(): string {
-    return trimTrailingSlashes(this.prootFilepath);
+    this.baseRequestURI = trimTrailingSlashes(base);
+    this.rootFilepath = trimTrailingSlashes(normalizePath(rootFilepath));
+    this.types = { ...mime.types, ...overrideTypes };
   }
 
   /**
-   * Strips the baseRequestURI from the identifier and checks if the stripped base URI matches the store's one.
-   * @param identifier - Incoming identifier.
+   * Maps the given resource identifier / URL to a file path.
+   * Determines the content-type if no content-type was provided.
+   * For containers the content-type input gets ignored.
+   * @param identifier - The input identifier.
+   * @param contentType - The (optional) content-type of the resource.
    *
-   * @throws {@link NotFoundHttpError}
-   * If the identifier does not match the baseRequestURI path of the store.
-   *
-   * @returns Absolute path of the file.
+   * @returns A ResourceLink with all the necessary metadata.
    */
-  public mapUrlToFilePath(identifier: ResourceIdentifier, id = ''): string {
-    return this.getAbsolutePath(this.getRelativePath(identifier), id);
+  public async mapUrlToFilePath(identifier: ResourceIdentifier, contentType?: string): Promise<ResourceLink> {
+    let path = this.getRelativePath(identifier);
+
+    if (path.includes('/..')) {
+      throw new UnsupportedHttpError('Disallowed /.. segment in URL.');
+    }
+
+    path = this.getAbsolutePath(path);
+
+    // Container
+    if (identifier.path.endsWith('/')) {
+      return {
+        identifier,
+        filePath: path,
+      };
+    }
+
+    // Would conflict with how new extensions get stored
+    if (/\$\.\w+$/u.test(path)) {
+      throw new UnsupportedHttpError('Identifiers should not match /\\$\\.\\w+$/.');
+    }
+
+    // Existing file
+    if (!contentType) {
+      const name = path.slice(path.lastIndexOf('/') + 1);
+      const folder = path.slice(0, path.length - name.length);
```

Comment:
```suggestion
      const [, name, folder] = path.match(/^(.*\/)(.*)$/)
```
Also, might want an `assert` earlier up that there is indeed a slash in the path.

Reply:
Update: assert for "starts with slash".